src/heap.rs

```diff
@@ -20,71 +18,130 @@ impl HeapId {
+        let end_page = capacity.saturating_sub(1) >> 12;
+        let page_count = end_page + 1;
+        let new_len = recycled_pages.len().saturating_sub(page_count);
+        let recycled_pages = recycled_pages.drain(new_len..).map(|mut page| {
+            page.0.fill(0);
+            Some(page)
+        });
+        let pages = recycled_pages
+            .chain(iter::repeat_with(|| Some(HeapPage::default())))
+            .take(page_count)
+            .collect();
+        Self { pages }
+    }
+
+    fn page(&self, idx: usize) -> Option<&HeapPage> {
+        self.pages.get(idx)?.as_ref()
+    }
+
+    fn get_or_insert_page(
+        &mut self,
+        idx: usize,
+        recycled_pages: &mut Vec<HeapPage>,
+    ) -> &mut HeapPage {
+        if self.pages.len() <= idx {
+            self.pages.resize(idx + 1, None);
         }
+        self.pages[idx].get_or_insert_with(|| recycled_pages.pop().unwrap_or_default())
+    }
+
+    fn write_u256(&mut self, start_address: u32, value: U256, recycled_pages: &mut Vec<HeapPage>) {
+        let mut bytes = [0; 32];
+        value.to_big_endian(&mut bytes);
 
-        value.to_big_endian(&mut self.0[start_address as usize..end]);
+        let offset = start_address as usize;
+        let (page_idx, offset_in_page) = (offset >> 12, offset & (HEAP_PAGE_SIZE - 1));
+        let len_in_page = 32.min(HEAP_PAGE_SIZE - offset_in_page);
+        let page = self.get_or_insert_page(page_idx, recycled_pages);
+        page.0[offset_in_page..(offset_in_page + len_in_page)]
+            .copy_from_slice(&bytes[..len_in_page]);
+
+        if len_in_page < 32 {
+            let page = self.get_or_insert_page(page_idx + 1, recycled_pages);
+            page.0[..32 - len_in_page].copy_from_slice(&bytes[len_in_page..]);
+        }
     }
 }
 
 impl HeapInterface for Heap {
     fn read_u256(&self, start_address: u32) -> U256 {
         self.read_u256_partially(start_address..start_address + 32)
     }
+
     fn read_u256_partially(&self, range: Range<u32>) -> U256 {
-        let end = (range.end as usize).min(self.0.len());
+        let offset = range.start as usize;
+        let (page_idx, offset_in_page) = (offset >> 12, offset & (HEAP_PAGE_SIZE - 1));
+        let mut result = [0_u8; 32];
+        let len_in_page = range.len().min(HEAP_PAGE_SIZE - offset_in_page);
+        if let Some(page) = self.page(page_idx) {
+            result[0..len_in_page]
+                .copy_from_slice(&page.0[offset_in_page..(offset_in_page + len_in_page)]);
+        }
 
-        let mut bytes = [0; 32];
-        for (i, j) in (range.start as usize..end).enumerate() {
-            bytes[i] = self.0[j];
+        if len_in_page < range.len() {
+            if let Some(page) = self.page(page_idx + 1) {
+                result[len_in_page..range.len()]
+                    .copy_from_slice(&page.0[..range.len() - len_in_page]);
+            }
         }
-        U256::from_big_endian(&bytes)
+        U256::from_big_endian(&result)
```

Comment:
We should definitely store the data in little-endian byteorder. In the Vec-based version that would've required rewriting Vec, but here it is relatively trivial. Just reverse the indices of pages.